src/vec_int128_ppc.h

```diff
@@ -1827,403 +1829,220 @@ vec_msumudm (vui64_t a, vui64_t b, vui128_t c)
+ *  @return vector of the high order 64-bits of the signed product
+ *  of the doubleword elements from vra and vrb.
+ */
+static inline vui64_t
+vec_mulhud (vui64_t vra, vui64_t vrb)
+{
+  return vec_mrgahd (vec_vmuleud (vra, vrb), vec_vmuloud (vra, vrb));
 }
 
 /** \brief Vector multiply odd unsigned doublewords.
  *
- *  Multiple the odd 64-bit doublewords of two vector unsigned long values and return
- *  the unsigned __int128 product of the odd doublewords.
+ *  Multiple the odd 64-bit doublewords of two vector unsigned
+ *  long values and return the unsigned __int128 product of the odd
+ *  doublewords.
+ *
+ *  \note The element numbering changes between big and little-endian.
+ *  So the compiler and this implementation adjusts the generated code
+ *  to reflect this.
  *
  *  |processor|Latency|Throughput|
  *  |--------:|:-----:|:---------|
  *  |power8   | 21-23 | 1/cycle  |
- *  |power9   | 5-7   | 2/cycle  |
+ *  |power9   | 8-13  | 2/cycle  |
  *
  *  @param a 128-bit vector unsigned long.
  *  @param b 128-bit vector unsigned long.
- * @ return vector unsigned __int128 product of the odd double words of a and b.
+ *  @ return vector unsigned __int128 product of the odd double words
+ *  of a and b.
  */
 static inline vui128_t
 vec_muloud (vui64_t a, vui64_t b)
 {
-  vui64_t res;
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  return vec_vmuleud (a, b);
+#else
+  return vec_vmuloud (a, b);
+#endif
+}
 
+/** \brief Vector Multiply Unsigned Doubleword Modulo.
+ *
+ *  Multiple the corresponding doubleword elements of two vector
+ *  unsigned long values and return the low order 64-bits of the
+ *  128-bit product for each element.
+ *
+ *  \note vec_muludm can be used for unsigned or signed integers.
+ *  It is the vector equivalent of Multiply Low Doubleword.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   | 19-28 | 1/cycle  |
+ *  |power9   | 11-16 | 1/cycle  |
+ *
+ *  @param vra 128-bit vector unsigned long.
+ *  @param vrb 128-bit vector unsigned long.
+ *  @return vector of the low order 64-bits of the unsigned product
+ *  of the doubleword elements from vra and vrb.
+ */
+static inline vui64_t
+vec_muludm (vui64_t vra, vui64_t vrb)
+{
 #ifdef _ARCH_PWR9
-  const vui64_t zero = { 0, 0 };
-  vui64_t b_oud = vec_mergel (zero, b);
-  __asm__(
-      "vmsumudm %0,%1,%2,%3;\n"
-      : "=v" (res)
-      : "v" (a), "v" (b_oud), "v" (zero)
-      : );
+  return vec_mrgald (vec_vmuleud (vra, vrb), vec_vmuloud (vra, vrb));
 #else
 #ifdef _ARCH_PWR8
-  const vui64_t zero = { 0, 0 };
-  vui64_t p0, p1, pp10, pp01;
-  vui32_t m0, m1;
-
-  m0 = vec_mergel ((vui32_t) b, (vui32_t) b);
-  m1 = (vui32_t) vec_splat ((vui64_t) a, 1);
-  p0 = vec_muleuw (m1, m0);
-  p1 = vec_mulouw (m1, m0);
+  vui64_t s32 = { 32, 32 }; // shift / rotate amount.
+  vui64_t z = { 0, 0 };
+  vui64_t t2, t3, t4;
+  vui32_t t1;
 
-  /* res[1] = p1[1];  res[0] = p0[0];  */
+  t1 = (vui32_t)vec_vrld (vrb, s32);
```

Comment:
_nit_ space after cast
_global_

Reply:
Ok, a lot of changes. But you asked for it.